### PR TITLE
add support for paged LDAP queries

### PR DESCRIPTION
**Pull request by stevekuznetsov**
/cc @liggitt

### CODE REVIEW

File: pkg/auth/ldaputil/query.go [hunk 1]
```diff
@@ -50,10 +51,17 @@ type LDAPQuery struct {
 
 	// Filter is a valid LDAP search filter that retrieves all relevant entries from the LDAP server with the base DN
 	Filter string
+
+	// PageSize is the maxmimum preferred page size, measured in LDAP entries
```
Comment (liggitt) [inline pkg/auth/ldaputil/query.go]: doc that values <= 0 mean no paging will be done

File: pkg/cmd/server/api/types.go [hunk 2]
```diff
@@ -1004,6 +1004,9 @@ type LDAPQuery struct {
 
 	// Filter is a valid LDAP search filter that retrieves all relevant entries from the LDAP server with the base DN
 	Filter string
+
+	// PageSize is the maxmimum preferred page size, measured in LDAP entries
+	PageSize int
```
Comment (deads2k) [inline pkg/cmd/server/api/types.go]: Do we really think that someone will set this?   Why not choose a number between 100 and 1000 and call it a day?

Comment (liggitt) [inline pkg/cmd/server/api/types.go]: some servers might not support paging, might not allow the particular user to page, or might have min/max sizes incompatible with what we pick

**Comment by stevekuznetsov** [comment]
[test][extended:ldap_groups]

**Comment by stevekuznetsov** [comment]
@liggitt ready for review

**Comment by stevekuznetsov** [comment]
docs PR: https://github.com/openshift/openshift-docs/pull/1591

File: pkg/auth/ldaputil/testclient/testclient.go [hunk 3]
```diff
@@ -143,3 +143,23 @@ func (c *DNMappingClient) Search(searchRequest *ldap.SearchRequest) (*ldap.Searc
 
 	return c.Client.Search(searchRequest)
 }
+
+// NewPagingOnlyClient returns a new PagingOnlyClient sitting on top of the parent client. This client returns the
+// provided search response for any calls to SearchWithPaging, or defers to the parent if the call is not to the
+// paged search function.
+func NewPagingOnlyClient(parent ldap.Client, response *ldap.SearchResult) ldap.Client {
```
Comment (liggitt) [inline pkg/auth/ldaputil/testclient/testclient.go]: why not just have a new var in Fake client named SearchWithPagingResponse?

Comment (stevekuznetsov) [inline pkg/auth/ldaputil/testclient/testclient.go]: These clients are made to be layered and this allows for the minimal amount of changes for insertion? I don't feel strongly.

**Comment by liggitt** [comment]
fun:
```
The following UPSTREAM commits have invalid summaries:
  [1c16be4] UPSTREAM: gopkg.in/ldap.v2: 51: exposed better API for paged search
```
@ironcladlou

**Comment by stevekuznetsov** [comment]
Better feedback on what is wrong would be great with that tool

**Comment by liggitt** [comment]
I think the feedback is as clear as it could possibly be. The regex was just wrong

**Comment by stevekuznetsov** [comment]
I didn't even see the feedback at the very bottom, on the same line as the make failure output!

**Comment by liggitt** [comment]
update UPSTREAM commit to bump to 07a7330929b9ee80495c88a4439657d89c7dbd87

File: pkg/auth/ldaputil/query.go [hunk 4]
```diff
@@ -137,6 +149,11 @@ func (o *LDAPQueryOnAttribute) buildAttributeQuery(attributeValue string,
 
 	filter := fmt.Sprintf("(&(%s)(%s))", o.Filter, specificFilter)
```
Comment (liggitt) [inline pkg/auth/ldaputil/query.go]: unrelated, but just noticed this... are we double parenthesizing the `o.Filter` incorrectly?

File: test/extended/ldap_groups.sh [hunk 5]
```diff
@@ -96,7 +96,7 @@ LDAP_SERVICE_IP=$(oc get --output-version=v1beta3 --template="{{ .spec.portalIP
 
 function compare_and_cleanup() {
 	validation_file=$1
-	actual_file=actual-${validation_file}.yaml
+	actual_file=actual-${validation_file}
```
Comment (liggitt) [inline test/extended/ldap_groups.sh]: why didn't we want this to be yaml?

Comment (liggitt) [inline test/extended/ldap_groups.sh]: nm, saw they all pass suffixed values already

File: pkg/auth/ldaputil/query.go [hunk 6]
```diff
@@ -50,10 +51,17 @@ type LDAPQuery struct {
 
 	// Filter is a valid LDAP search filter that retrieves all relevant entries from the LDAP server with the base DN
 	Filter string
+
+	// PageSize is the maxmimum preferred page size, measured in LDAP entries. A page size of 0 means no paging will be done.
```
Comment (liggitt) [inline pkg/auth/ldaputil/query.go]: s/maxmimum/maximum/

**Comment by liggitt** [comment]
nits, LGTM after updating go-ldap

File: pkg/auth/ldaputil/query.go (hunk 4 above)
Comment (stevekuznetsov) [inline pkg/auth/ldaputil/query.go]: It seems like we are, but the double-parens are tolerated? Our test sync configs in extended test declare filters with parens and they run fine.

**Comment by liggitt** [comment]
`bump(...): ...`

File: Godeps/Godeps.json [hunk 7]
```diff
@@ -1037,7 +1037,7 @@
 		{
 			"ImportPath": "gopkg.in/ldap.v2",
 			"Comment": "v2.2",
```
Comment (liggitt) [inline Godeps/Godeps.json]: update comment to 2.2.1

**Comment by stevekuznetsov** [comment]
why is that a different syntax.

File: pkg/auth/ldaputil/query.go (hunk 4 above)
Comment (liggitt) [inline pkg/auth/ldaputil/query.go]: tolerated, but unnecessary. I just wondered

**Comment by liggitt** [comment]
need to update godeps in the bump commit... the commitchecker will complain at you otherwise

**Comment by liggitt** [comment]
LGTM, [merge]

**Comment by stevekuznetsov** [comment]
@liggitt please re-tag

**Comment by liggitt** [comment]
[merge]

**Pull request merged**